### ISSUE TITLE
epoll reactor stdio fails when output redirected to a file

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4429]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4429|
|Type|defect|
|Created|2010-05-11 13:59:43Z|
|Branch|https://github.com/twisted/twisted/tree/epoll-EPERM-4429|

This is like [[#2259](https://github.com/twisted/twisted/issues/2259)](https://github.com/twisted/twisted/issues/2259), but only for epollreactor.  #2259 fixed the behavior for all the other reactors.

<details><summary>Searchable metadata</summary>

```
trac-id__4429 4429
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__Twisted_12_1 Twisted-12.1
branch__branches_epoll_EPERM_4429 branches/epoll-EPERM-4429
branch_author__itamar itamar
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1273586383000000 1273586383000000
changetime__1336157393000000 1336157393000000
version__None None
owner__itamar itamar
cc__itamarst cc__nekto0n
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Implementing this with immediate writes that are assumed to not block, and reads driven by coiterate(), may be sufficient for a first pass; producer API can be done as separate ticket.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [34301]) Branching to 'epoll-EPERM-4429'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst removed owner|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/epoll-EPERM-4429

Design is basically as described above in comment 3.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_7"></a>|@jyknight commented|
|-|-|

Why a 0.001sec delay rather than 0sec? Other reactors will trigger file reads/writes every iteration.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Thanks.

  1. `twisted.internet.epollreactor` isn't necessarily importable.  New tests need to handle the import error and skip as necessary.
  1. `EpollEPERMTests` creates a reactor and leaks a bunch of its state.  See `ReactorBuilder` for all the things you have to do to clean up after a reactor.  It seems like most or all of the tests on this class actually should be `ReactorBuilder` tests and moved to a general stdio test module.  The functionality should work on all reactors, not just epoll, shouldn't it?
  1. To stick with the theme foom raised, the Twisted Committee Against Arbitrary Magic Numbers recommends a delay of 0 instead of 0.001.
  1. `PosixReactorBase._disconnectSelectable` seems to have a lot more logic than `_ContinuousPolling._disconnectSelectable`.  Does this stdio implementation still support half close?  Does it clean up the readers and writers properly?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst removed owner|
|-|-|

Review comments were all addressed; new build:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/epoll-EPERM-4429

I'm pretty sure the documentation buildslave's complaints are bogus.

**Comment by twisted-trac** [comment]
|<img alt="nekto0n's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|nekto0n commented|
|-|-|

I think [this](http://twistedmatrix.com/trac/browser/trunk/twisted/internet/_posixstdio.py#L47) comment should fixed too.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

It is fixed, in the branch under review. You can see the branch in the ticket description at the top of the page.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

 * pyflakes:
```
twisted/internet/test/test_stdio.py:52: local variable 'stdio' is assigned to but never used
twisted/internet/test/test_stdio.py:76: local variable 'stdio' is assigned to but never used
twisted/internet/test/test_epollreactor.py:10: 'EPollReactor' imported but unused
twisted/internet/_posixstdio.py:14: 'errno' imported but unused
twisted/internet/_posixstdio.py:14: 'os' imported but unused
```

 * 
```
+                self._loop.start(0.00000001, now=False)
```
 Maybe it's worth reusing task._EPSILON?

 * 
```
+            if e.errno == errno.EPERM:
+                self._continuousPolling.addReader(reader)
```
 Can you add a small comment around those exceptions?

 * There are some chars after 80 in test_epollreactor.py
 *
```
        self.extraFile = file(path, "r+")
```
 Shouldn't we close this in tearDown?

Some small questions, but the branch seems fine, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst commented|
|-|-|

Closing file in tearDown may cause problem storing extraFile as attribute was supposed to fix, where reactor cleanup tries to operate on closed file descriptor. So I'm not going to bother. Everything else is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst set status to `closed`|
|-|-|

(In [34346]) Merge epoll-EPERM-4429: epoll reactor now supports filesystem files about as well as other reactors do.

Author: itamar
Review: exarkun, jknight, therve
Fixes: [#4429](https://github.com/twisted/twisted/issues/4429)

If you do redirect stdin or stdout to a file, epoll reactor will no longer blow up.
